datadog_checks_base/datadog_checks/base/checks/windows/perf_counters/counter.py

```diff
@@ -0,0 +1,516 @@
+
+        if not handles_with_data:
+            return
+
+        if self.aggregate is not False:
+            if self.average:
+                self.aggregate_transformer(total / handles_with_data, tags=self.tags)
+            else:
+                self.aggregate_transformer(total, tags=self.tags)
+
+    def refresh(self, instance_counts):
+        old_instances = self.instances
+        new_instances = {}
+
+        for instance, current_count in instance_counts.items():
+            if instance in old_instances:
+                counter_handles = old_instances.pop(instance)
+                new_instances[instance] = counter_handles
+                old_count = len(counter_handles)
+
+                if current_count > old_count:
+                    for index in range(old_count, current_count):
+                        path = construct_counter_path(
+                            machine_name=self.connection.server,
+                            object_name=self.object_name,
+                            counter_name=self.name,
+                            instance_name=instance,
+                            instance_index=index,
+                        )
+                        counter_handle = self.counter_selector(self.connection.query_handle, path)
+                        counter_handles.append(counter_handle)
+                elif current_count < old_count:
+                    for _ in range(old_count - current_count):
+                        counter_handle = counter_handles.pop()
+
+                        # https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhremovecounter
+                        # http://timgolden.me.uk/pywin32-docs/win32pdh__RemoveCounter_meth.html
+                        win32pdh.RemoveCounter(counter_handle)
+            else:
+                counter_handles = []
+                new_instances[instance] = counter_handles
+
+                for index in range(current_count):
+                    path = construct_counter_path(
+                        machine_name=self.connection.server,
+                        object_name=self.object_name,
+                        counter_name=self.name,
+                        instance_name=instance,
+                        instance_index=index,
+                    )
+                    counter_handle = self.counter_selector(self.connection.query_handle, path)
+                    counter_handles.append(counter_handle)
+
+        # Remove expired instances
+        self.clear()
+
+        self.instances = new_instances
+
+    def clear(self):
+        for counter_handles in self.instances.values():
+            while counter_handles:
+                counter_handle = counter_handles.pop()
+
+                # https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhremovecounter
+                # http://timgolden.me.uk/pywin32-docs/win32pdh__RemoveCounter_meth.html
+                win32pdh.RemoveCounter(counter_handle)
```

Comment:
do you think this can ever return the errors [here](https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhremovecounter#return-value)?

Reply:
perhaps, addressed!